File: 3. Longest Substring Without Repeating Characters.py
# Letcode problem:https://leetcode.com/problems/longest-substring-without-repeating-characters/
# Level: medium
# Given a string s, find the length of the longest substring without repeating characters.

class Solution:
    def max_non_repitive(self, sub_list):
        buff_list = {}
        if len(sub_list) == 1:
            return 1
        for item in sub_list:
            if item not in buff_list:
                buff_list[item] = item
            else:
                break
        return len(buff_list)

    def lengthOfLongestSubstring(self, s: str) -> int:
        lenght = 0
        str_list = [char for char in s]
        print("s:", str_list)
        for idx in range(len(str_list)):
            for list_lengh in range(idx+lenght-1, len(str_list)):
                list_lengh = list_lengh+1
                sub_list = str_list[idx:list_lengh]
                if len(sub_list) > lenght:
                    length_temp = self.max_non_repitive(sub_list)
                    print("sub_list[{}:{}]:{}:  max2:{}".format(
                        idx, list_lengh, sub_list, length_temp))
                    if lenght < length_temp:
                        lenght = length_temp
                    if length_temp < len(sub_list):
                        break
        return lenght

    def lengthOfLongestSubstring2(self, s: str) -> int:
        lenght = 0
        str_list = [char for char in s]
        print("s:", str_list)
        for idx in range(len(str_list)):
            sub_list = str_list[idx:]
            if len(sub_list) <= lenght:
                break
            length_temp = self.max_non_repitive(sub_list)
            print("sub_list[{}:]:{}:  max:{}".format(
                idx, sub_list, length_temp))
            if lenght < length_temp:
                lenght = length_temp
        return lenght


if __name__ == '__main__':
    solution = Solution()
    # s = "pwwkew"
    s = "abcabcbb"
    # s = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    print("String len:", len(s))
    print(solution.lengthOfLongestSubstring(s))
    # lst = [50, 70, 30, 20, 90, 10, 50]
    # print(lst[1:3])

File: 3. Longest Substring Without Repeating Characters.go
// Author: Le Anh Tuan (latuannetnam@gmail.com)
// 3. Longest Substring Without Repeating Characters
// Letcode problem: Given a string s, find the length of the longest substring without repeating characters.
// Level: medium

// Input: s = "abcabcbb"
// Output: 3
// Explanation: The answer is "abc", with the length of 3.

package main

import "fmt"

// Use sliding window
// Runtime: 7 ms, faster than 77.54% of Go online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 2.7 MB, less than 75.59% of Go online submissions for Longest Substring Without Repeating Characters.
func lengthOfLongestSubstring(s string) int {
	max_lenght := 0
	start := 0
	end := 0
	char_hash := map[byte]bool{}
	// Start window with size = 0
	for end < len(s) {
		if !char_hash[s[end]] {
			char_hash[s[end]] = true
			// Increase windows size
			end++
			// Store window size and compare with max length
			window_size := end - start
			if max_lenght < window_size {
				max_lenght = window_size
			}
			fmt.Printf("substr: %s max len:%d\n", s[start:end], max_lenght)
		} else {
			// Repeated char
			// Remove char_hash[s[start]]
			char_hash[s[start]] = false
			// Slide window
			start++
		}
	}
	return max_lenght
}

// Best solution
func lengthOfLongestSubstringBest(s string) int {
	location := [256]int{}

	for i := range location {
		location[i] = -1
	}

	left, maxLen := 0, 0
	for i := 0; i < len(s); i++ {
		if location[s[i]] >= left {
			left = location[s[i]] + 1
		} else if i+1-left > maxLen {
			maxLen = i + 1 - left
		}
		location[s[i]] = i
	}
	return maxLen
}
func mainLengthOfLongestSubstring() {
	s := "abcabcbb"
	fmt.Printf("string:%s\n", s)
	fmt.Printf("max len:%d\n", lengthOfLongestSubstring(s))

	s = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	fmt.Printf("string:%s\n", s)
	fmt.Printf("max len:%d\n", lengthOfLongestSubstring(s))
}
